### DISCUSSION TITLE
Parallel findOneOrFail do not utilize identity map

### DISCUSSION THREAD

**Discussion by ghempton** [mikro-orm/mikro-orm#5107] [General]
### Describe the bug

We are noticing that multiple `findOneOrFail` calls with the same id that are run in parallel seem to unexpectedly run multiple queries. The expected behavior would be for one a single query to be ran. This becomes a problem when resolving graphql heirarchies in nestjs, where the same entity could be resolved many times in a single unit of work.

### Reproduction

```
const first = this.em.findOneOrFail(Agent, id);
const second = this.em.findOneOrFail(Agent, id);
const res = await Promise.all([first, second]);
```

Results in the following queries in debug mode:

```
[query] select "a0".* from "agent" as "a0" where "a0"."id" = '4ff88985-93aa-4fe6-b49c-bcb32338dc95' limit 1 [took 4 ms, 1 result]
[query] select "a0".* from "agent" as "a0" where "a0"."id" = '4ff88985-93aa-4fe6-b49c-bcb32338dc95' limit 1 [took 9 ms, 1 result]
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.1

### Node.js version

v18.12.0

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> The expected behavior would be for one a single query to be ran

There is no dataloader behind `em.findOne`, those queries are not supposed to be batched. Identity map can help only if some call finishes first.

@darkbasic is working on https://github.com/darkbasic/mikro-orm-dataloaders which I believe should help with this, but looks like it currently only supports `em.find` and not `em.findOne`

**Reply by darkbasic** [reply]
It supports both find and findOne. Please note that it's still considered experimental and the API will probably change soon so use the new custom entity manager instead of the custom repository.

**Reply by darkbasic** [reply]
Also it doesn't cover the full set of operators yet.

**Reply by ghempton** [reply]
Thanks for the fast response! How does that project relate to https://mikro-orm.io/docs/dataloaders? I was under the impression that dataloaders were going to be part of core in v6?

**Reply by ghempton** [reply]
I now see that data loaders only apply to .load(). I think we may be able to avoid findOne* api calls and use load() to solve this in our use case.

**Reply by darkbasic** [reply]
> I think we may be able to avoid findOne* api calls and use load() to solve this in our use case.

Yeah I think so, at least in my use case I've been able to get the job done with references and collections (you can dataload both) 90% of the times. Let me know how it works for your use case!
